crates/debug/src/transform/attr.rs

```diff
@@ -52,11 +54,6 @@ where
 
     let range_info = if let Some(subprogram_range_builder) = subprogram_range_builder {
         subprogram_range_builder
-    } else if entry.tag() == gimli::DW_TAG_compile_unit {
-        // FIXME currently address_transform operate on a single func range,
```

Comment:
I don't think removing it fixes the issue? I would like to see some guaranties that all ranges are transformed, e.g. range that spans multiple functions (IIRC `RangeInfoBuilder`/`AddressTransform` can do only inner function transforms)

Reply:
From my `moc` test:
Input
```
              DW_AT_ranges [DW_FORM_sec_offset]	(0x0000000c
                 [0x00000007, 0x00000013)
                 [0x00000018, 0x00000038)
                 [0x0000003d, 0x00000067)
                 [0x0000006c, 0x00000081)
                 [0x00000086, 0x000000df)
                 [0x000000e4, 0x000000ed)
...
```
Output by `wasm2obj`:
```
0x0000000b: DW_TAG_compile_unit [1] *
              DW_AT_ranges [DW_FORM_sec_offset]	(0x00000000
                 [0x000000000000011c, 0x000000000000011d) ".text._wasm_function_21"
                 [0x0000000000000000, 0x0000000000000032)
                 [0x0000000000000000, 0x0000000000000052) ".text._wasm_function_1"
                 [0x0000000000000000, 0x00000000000000cd) ".text._wasm_function_2"
                 [0x0000000000000000, 0x0000000000000041) ".text._wasm_function_3"
                 [0x0000000000000000, 0x0000000000000040) ".text._wasm_function_4"
                 [0x0000000000000000, 0x0000000000000034) ".text._wasm_function_5"
                 [0x0000000000000000, 0x0000000000000031) ".text._wasm_function_6"
... // pretty consistently, but then
                 [0x0000000000000000, 0x000000000000007a) ".text._wasm_function_179"
                 [0x000000000000008c, 0x000000000000027b) ".text._wasm_function_179"
                 [0x0000000000000000, 0x000000000000009c) ".text._wasm_function_180"
                 [0x000000000000009c, 0x0000000000000197) ".text._wasm_function_180"
                 [0x0000000000000000, 0x000000000000007c) ".text._wasm_function_181"
                 [0x0000000000000092, 0x0000000000000097) ".text._wasm_function_181"
                 [0x0000000000000000, 0x000000000000010f) ".text._wasm_function_182"
                 [0x000000000000010f, 0x0000000000000200) ".text._wasm_function_182"
                 [0x0000000000000000, 0x0000000000000057) ".text._wasm_function_183"
                 [0x0000000000000000, 0x00000000000000a4) ".text._wasm_function_184"
...
```
I am not really in the position to understand what the first column means (start address?) and why it is mostly zeros, but not always. I assume the second column is length. The first two lines look garbled, though. We could make this a test (after the wrinkles are ironed out), but which tools would generate `DW_AT_ranges`? Maybe `rustc` (which is LLVM-based).

Also, will the backend (`cranelift`) be confused by the DWARF-4 output of `wasmtime` containing `DW_AT_ranges`?

Reply:
Right, that's what I would expect: ranges are sub-ranges of a single function. In object file first number will describe offset from the symbol start, so the above output somewhat expected. A gap in ".text._wasm_function_179" ranges may signify that there is some constant pool in the middle of the function ? The change code in question is what if a tool outputs CU with single range that covers all functions.

Reply:
I'm okay with removing it, only if there is a guarantee, at RangeInfoBuilder/AddressTransform level, that a range that covers multiple functions, by some tool, will not be wrongfully converted.

Reply:
Just a remark, after commit cc1e974 now some dirt seems cleaned up, and I see:
```
0x0000000b: DW_TAG_compile_unit [1] *
              DW_AT_ranges [DW_FORM_sec_offset]	(0x00000000
                 [0x0000000000000000, 0x0000000000000032)
                 [0x0000000000000000, 0x000000000000008c) ".text._wasm_function_1"
                 [0x0000000000000000, 0x00000000000000cd) ".text._wasm_function_2"
                 [0x0000000000000000, 0x000000000000008c) ".text._wasm_function_3"
                 [0x0000000000000000, 0x0000000000000168) ".text._wasm_function_4"
                 [0x0000000000000000, 0x0000000000000034) ".text._wasm_function_5"
```
the first row still makes me suspicious...

Reply:
Yeah, interesting to know where it is coming from. Maybe it is some memory addresses?

Reply:
It was the discrepancy between `DebugAddrIndex(0)`, the initialised, wrong value, and `DebugAddrIndex(8)`, the read, correct value. Now both are in sync.

Reply:
It should be possible to find out by gripping for` addrx` and `0x00`.

Reply:
I looked into the `RangeInfoBuilder::build()` code. It seems to straightforwardly map each range through the address translator. Then it writes the result obtaining an ID and passes that on. In the DWARF-4 output there is a `.debug_range` section, which seems legit. Since this is internally consumed by `wasmtime`, I don't see a lot of risk. When using `wasm2obj`, it remains to see whether downstream toolchains are happy with this. But I think this use case would be an extreme minority, especially when it comes to DWARF debugging. Even then it should be a matter of a few lines of code to detect that we are outputting DWARF-4 and in this case just emit a `(DW_AT_low_pc, DW_AT_high_pc)` pair.
Unfortunately I cannot offer other guarantees besides my offer to fix things in case some tool breaks because of this. What do you think?

Reply:
> It seems to straightforwardly map each range through the address translator. 

Correct. Concern about what if the range spans all functions (not per functions), e.g. CU range is 0..100 but individual functions are 0..20, 0..50, etc.

Let's change the `else if`, but keep the comment.